### PR TITLE
Add TTL to all lightblue locks by way of LockStrategy

### PR DESCRIPTION
**Pull request by alechenninger**
Abstracting out this concept and injecting it also makes it much easier to test concurrent scenarios.

~~Still WIP:~~

- [x] Using LockStrategy in NotificationRepository also
- [x] More tests

### CODE REVIEW

**Comment by alechenninger** [comment]
Ready for review @bserdar @kahowell

**Comment by alechenninger** [comment]
I will add some tests specifically for the lightblue lock strategy, but since this likely won't change much, should still be good to review, just wait to merge until I add those tests and those are reviewed

**Comment by alechenninger** [comment]
Tests done

File: lib/src/main/java/org/esbtools/eventhandler/PollingDocumentEventProcessorRoute.java [hunk 1]
```diff
@@ -50,7 +53,7 @@ public PollingDocumentEventProcessorRoute(DocumentEventRepository documentEventR
     @Override
     public void configure() throws Exception {
         from("timer:pollForDocumentEvents?period=" + pollingInterval.toMillis())
-        .routeId("documentEventsProcessor")
+        .routeId("documentEventsProcessor-" + idCounter.getAndIncrement())
```
Comment (alechenninger) [inline lib/src/main/java/org/esbtools/eventhandler/PollingDocumentEventProcessorRoute.java]: This and below incrementing id to routeId allows a single camel context to have multiple of these routes, which is useful among other things for testing concurrent scenarios.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+ * is released.
+ *
+ * <p>Lightblue locks without TTLs are unsafe. If a client crashes, the lock will not be released.
+ * However, locks with TTLs may expire prematurely. By pinging the lock in a separate thread
+ * periodically, we can substantially lesson the likelihood of unintentionally losing the lock.
+ *
+ * <p>Consumers are expected to check the lock at critical points to ensure it did not expire by
+ * calling {@link LockedResource#ensureAcquiredOrThrow(String)}.
+ */
+public class LightblueAutoPingLockStrategy implements LockStrategy {
```
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java]: This is somewhat redundant with lightblue client's interface, however, as LockStrategy doc should explain some, having this here gives us flexibility to use something else for locking, which is necessary for sane testing strategies and helps pull out some of the non-trivial configuration of such a strategy out of the repositories' already large constructors. And obviously we naturally get to implement the auto-pinging stuff here.

File: lightblue/src/test/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepositoryTest.java [hunk 3]
```diff
@@ -189,15 +202,16 @@ public void shouldReturnNonOverlappingSetsEvenAmongMultipleThreads() throws Ligh
 
             Future<List<LightblueDocumentEvent>> futureThread1Events = executor.submit(() -> {
                 bothThreadsStarted.countDown();
-                return thread1Repository.retrievePriorityDocumentEventsUpTo(100);
+                return thread1Repository.retrievePriorityDocumentEventsUpTo(15);
             });
             Future<List<LightblueDocumentEvent>> futureThread2Events = executor.submit(() -> {
                 bothThreadsStarted.countDown();
-                return thread2Repository.retrievePriorityDocumentEventsUpTo(100);
+                return thread2Repository.retrievePriorityDocumentEventsUpTo(15);
```
Comment (alechenninger) [inline lightblue/src/test/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepositoryTest.java]: Tangentially related to core work of this PR; using lower numbers of events in this test makes it easier to understand when something goes wrong.

**Comment by alechenninger** [comment]
There is some additional locking that will be needed in the future for recovering processing events or notifications which failed to update to a processed status. I will do this in another PR.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LockedResource.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.esbtools.eventhandler.lightblue;
+
+import java.io.Closeable;
+
+public interface LockedResource extends Closeable {
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LockedResource.java]: Is this still required? There is a Locking.Lock class extending Closeable in client.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java [hunk 5]
```diff
@@ -80,10 +78,14 @@ public LightblueNotificationRepository(LightblueClient lightblue, String[] entit
             DataBulkRequest updateEntities = new DataBulkRequest();
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: A single update request with a query containing the _ids of docs to be updated would perform better here, instead of bulk request.

Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: I see. Eventually though I will be persisting different status updates based on the entity (processing or failed). I could turn that into two requests in a bulk request, I think though. Will open separate issue.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LockedResource.java (hunk 4 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LockedResource.java]: Semantics are a little different (throws with custom exception message instead of return false), plus I don't want to rely on lightblue client, in the client it's a class not an interface, and also we may have other impls in the future.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java (hunk 5 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: Added #11

Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: Similarly: in markNotificationsProcessedOrFailed: a bulk request with two
updates, one for processed, and one for failed, would perform better.

Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: Cool, I made the issue to look at all uses of bulk updates, but feel free to comment there if you find more. I'll copy this one. Ty

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java [hunk 6]
```diff
@@ -42,27 +42,25 @@
...
 
     @Override
     public List<LightblueNotification> retrieveOldestNotificationsUpTo(int maxEvents)
-            throws LightblueException {
-        try {
-            blockUntilLockAcquired(Locks.forNotificationsForEntities(entities));
-
+            throws Exception {
+        try (LockedResource lock = lockStrategy
+                .blockUntilAcquired(ResourceIds.forNotificationsForEntities(entities))) {
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: There is inherent assumption here that no entity can belong to two different notification repositories.

Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: This kind of locking is reducing potential parallelism. There can be only one active thread loading notifications with this implementation. A better approach might be to retrieve maxEvents*nNodes notifications, and randomly pick maxEvents of them, locking them one by one. That returns at most maxEvents notifications, but the algorithm runs on all nodes, with no blocking.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java [hunk 7]
```diff
@@ -42,27 +42,25 @@
...
             BulkLightblueRequester requester = new BulkLightblueRequester(lightblue);
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: One more thing to consider re: locking by entity name: if lock owner dies, all entity even processing stops until lock expires.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 8]
```diff
@@ -60,20 +59,19 @@
      * or superseded), should be put back in the document event entity pool to be processed later.
      */
     private final int documentEventBatchSize;
-    private final Locking locking;
+    private final LockStrategy lockStrategy;
     private final Map<String, ? extends DocumentEventFactory> documentEventFactoriesByType;
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Suggestion: DocumentEventFactories class with 
  DocumentEventFactory get(String type) throws NoSuchElementException, so we don't have to check everywhere

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 9]
```diff
@@ -153,18 +148,6 @@ public void markDocumentEventsProcessedOrFailed(
...
-                Thread.sleep(3000L);
-            } catch (InterruptedException e) {
-                // Just try again...
-            }
-        }
-    }
-
     /**
      * Creates {@link LightblueDocumentEvent}s from entities using
      * {@link #documentEventFactoriesByType}, optimizes away superseded and merge-able events, and
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Q: how useful is the parseAndOptimizeDocumentEventEntitiesUpTo()? It relies on the related events being in the same batch.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+     *                   {@code autoPingInterval} to ensure locks do not accidentally expire.
+     */
+    public LightblueAutoPingLockStrategy(Locking locking, Duration tryAcquireInterval,
+            Duration autoPingInterval, Duration timeToLive) {
+        this.locking = locking;
+        this.tryAcquireInterval = tryAcquireInterval;
+        this.autoPingInterval = autoPingInterval;
+        this.timeToLive = timeToLive;
+
+        if (timeToLive.compareTo(autoPingInterval) <= 0) {
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java]: Continuous pinging is not a good idea. TTL is there to ensure things end, or you lose them. This prevents that, and equivalent to setting ttl large.

A much better approach would be is to ping just before updating something, and then ping again, and if that fails, deal with cleanup.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java (hunk 6 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: We had talked about that initially, but I ran into some issues, at least for document event entities. I think we could make all of the synchronization more robust, but it's a question of what is appropriate for 1.0.0, given the time pressure? I'm more concerned with it working and reducing chance of lost / stuck events than performance right now. A matter of prioritization.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java (hunk 7 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: Correct. Looking at a TTL of around ~30 seconds or so, shouldn't be a huge concern IMO.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java (hunk 10 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java]: Counter arguments:

- Large TTL delays recovery (relates to your point about recovery per entity). There's not really a good way to determine a good TTL without background ping. Auto pinging in separate thread keeps TTL relatively low.
- It doesn't prevent losing when we want to lose it. We still release the lock and also stop the thread when doing so.
- If the node dies, so does the pinging thread, allowing lock to expire.

Am I missing something? I thought we talked about this kind of impl before.

In the future, we would like to not use lightblue for synchronizing distributed apps and instead look at something like ZooKeeper. I'd do this now, but this will add time.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 9 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Yes, hence locking per event type.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 8 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: I actually had a type like this, but Kevin thought this was more confusing that just using a good ol' Map. I agree. We only need to check in one place.

**Comment by alechenninger** [comment]
TODO:

- [ ] Refactor lock interfaces to better support updated usage / remove APIs we won't need
- [ ] Make notification repository use a somewhat similar strategy

There are some other error scenarios that are broken still, but I think I will fix them in another PR at this point b/c this is getting way too large.

File: lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java [hunk 11]
```diff
@@ -21,16 +21,35 @@
...
     /**
-     * Retrieves the oldest {@code maxEvents} {@link Notification notifications}, oldest first.
+     * Retrieves the oldest {@code maxNotifications} {@link Notification notifications}, oldest
+     * first.
      *
      * <p>Once retrieved, a notification should not be retrieved again, atomically. That is, many
-     * threads looking up notifications at the same time should all et a unique non-overlapping sample
-     * of the oldest notifications. Subsequent calls should always return a unique set.
+     * threads looking up notifications at the same time should all et a unique non-overlapping
+     * sample of the oldest notifications. Subsequent calls should always return a unique set.
```
Comment (bserdar) [inline lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java]: How hard of a requirement is it to retrieve "oldest" notifications? Are we talking about strict ordering, or preferential treatment? Randomized selections that prefer older notifications with higher likelihood will probably perform better than a strict ordering.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 12]
```diff
@@ -92,30 +101,77 @@ public void addNewDocumentEvents(Collection<? extends DocumentEvent> documentEve
...
+    /**
+     * {@inheritDoc}
+     *
+     * <p>N.B. This implementation currently works by simply checking if the known in memory
+     * timeouts of provided events fall within some threshold. It could be updated to also timestamp
+     * the persisted events again, but to be safe we would probably want to also lock them to do
+     * this update. For now, that complexity is probably not worth it.
+     */
+    @Override
+    public Collection<? extends DocumentEvent> checkExpired(Collection<? extends DocumentEvent> events) {
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Why is this necessary, especially if it is checking using events in memory, and using a lock clock? Some events this node might think not expired would be expired for some others, and vice versa.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 13]
```diff
@@ -92,30 +101,77 @@ public void addNewDocumentEvents(Collection<? extends DocumentEvent> documentEve
...
-                    .data(FindRequests.priorityDocumentEventsForEntitiesUpTo(entities, documentEventBatchSize))
-                    .parseProcessed(DocumentEventEntity[].class);
+        if (maxEvents == 0) {
+            return Collections.emptyList();
+        }
 
-            if (documentEventEntities.length == 0) {
-                return Collections.emptyList();
-            }
+        DocumentEventEntity[] documentEventEntities = lightblue
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: You retrieve a batch of event documents here, and then try to optimize by matching identities. The batch may not contain all events of a particular identity because of the batch size restriction. A better approach would be is to get a batch of identities, lock them, and then load all the documents that were acquired. Then you'll have all documents of a particular identity at a given moment.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 14]
```diff
@@ -147,69 +203,299 @@ public void markDocumentEventsProcessedOrFailed(
...
-     * than {@code maxEvents} will be returned. List may include events with newly computed entities
-     * that are not yet persisted.
+     * <p>Checks for lost locks before persisting, dropping and logging those lost.
      */
-    private List<LightblueDocumentEvent> parseAndOptimizeDocumentEventEntitiesUpTo(int maxEvents,
-            DocumentEventEntity[] documentEventEntities, List<DocumentEventEntity> entitiesToUpdate,
-            LightblueRequester requester) {
-        List<LightblueDocumentEvent> optimized = new ArrayList<>(maxEvents);
+    private List<LightblueDocumentEvent> persistNewEntitiesAndStatusUpdatesToExisting(
+            LockedResources<SharedIdentityEvents> identityLocks) throws LightblueException {
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: "NewEntities" is a bit misleading

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 15]
```diff
@@ -147,69 +203,299 @@ public void markDocumentEventsProcessedOrFailed(
...
+
+        List<SharedIdentityEvents> lostLocks = identityLocks.checkForLostResources();
+        List<LightblueDocumentEvent> savedEvents = new ArrayList<>();
+
+        // TODO: We make single request per event here (wrapped in bulk request). Maybe could optimize.
+        // Right now each event may have different processing date which we are looking for.
+        // Could probably change that so processing dates were more grouped.
+        // See: https://github.com/esbtools/event-handler/issues/11
+        for (SharedIdentityEvents lockedEvents : identityLocks.getResources()) {
+            if (lostLocks.contains(lockedEvents)) {
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Unnecessary complexity: there is no need to put lostLocks in a list, then perform a lookup. lockedEvents.isLocked() would be easier.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 16]
```diff
@@ -147,69 +203,299 @@ public void markDocumentEventsProcessedOrFailed(
...
+                continue;
+            }
+
+            for (DocumentEventUpdate update : lockedEvents.updates.values()) {
+                LightblueDocumentEvent event = update.event;
+
+                DocumentEventEntity entity = event.wrappedDocumentEventEntity();
+
+                if (entity.get_id() == null) {
+                    if (entity.getStatus().equals(DocumentEventEntity.Status.processing)) {
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Does "_id=null and status=processing" mean this is a new entity to be inserted because some events are merged? Maybe a comment here explaining that, or a more explicit way of identifying this situation is needed.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 17]
```diff
@@ -147,69 +203,299 @@ public void markDocumentEventsProcessedOrFailed(
...
+                        insertAndUpdateEvents.add(InsertRequests.documentEventsReturningOnlyIds(entity));
+
+                        savedEvents.add(event);
+                    }
+                } else {
+                    insertAndUpdateEvents.add(UpdateRequests.documentEventStatusIfCurrent(
+                            entity, update.originalProcessingDate));
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Instead of this, could use save with upsert=true?

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 18]
```diff
@@ -147,69 +203,299 @@ public void markDocumentEventsProcessedOrFailed(
...
+                                .map(e -> "Code: " + e.getErrorCode() + ", " +
+                                        "Context: " + e.getContext() + ", " +
+                                        "Message: " + e.getMsg())
+                                .collect(Collectors.toList());
+
+                        logger.warn("Event update failed. Will not process. Event was: <{}>. " +
+                                "Errors: <{}>", event, errorStrings);
+                    }
+
+                    eventsIterator.remove();
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Remove lock as well

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 15 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Came to same conclusion, +1. I'm removing the checkForLostResources API all together.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 13 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Good point, I'll think if this is easy to incorporate now, otherwise I'll open an issue for it.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 12 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: I think you should check back in before you publish documents.

This is the scenario I want to catch with this:

1. Start a transaction with identities
2. Take too long to build documents for those events and now those events are expired
3. If ignore the expiration, you'll publish duplicates. It's not the end of the world, but it can be prevented.

Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Does "events expired" mean locks expired? You should ping, update, and either unlock or ping again. If the second unlock/ping fails, then there's a failure, and you'll get a duplicate event, so you should deal with that.

File: lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java (hunk 11 above)
Comment (alechenninger) [inline lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java]: Yeah, it is a similar thing with priority document events. I want to do random but has to weight older notifications / higher priority document events. For now I am (or will be, in the case of notifications) just going down the list and locking in order.

Also, I wanted to ask for your thoughts on what to lock for notifications. We could do by identity like document events, which would open up opportunity to similarly check for superseded notifications, but currently we don't do this, and the APIs aren't there to do this. I'm leaning towards just locking by entity name or specific notification id for now and seeing how that goes.

Comment (bserdar) [inline lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java]: locking by entity name reduced parallelism, notification id should work better.

re: going down the list and locking: there might be a problem with contention. All nodes will do the same, over the same set of data. Retrieving a result set nnodes times the batch, and randomly locking batch items will parallelize better.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 18 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: We remove all locks after we return the document events, since they are marked processing with timestamp (this happens in the try-with-resources block)

File: lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java (hunk 11 above)
Comment (alechenninger) [inline lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java]: Do you think it would be an excessive amount of lock calls to lock for each notification id? Notification batches will typically be large I think; we should process notifications as fast as lightblue will allow us.

Comment (bserdar) [inline lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java]: There is no reason for batches to be large. Let this scale horizontally,
instead of trying to do a lot of things on a single node.

Lock calls are extremely lightweight *if* you acquire the lock the first
time you ask for it.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 12 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: It checks only timeouts, no locks. At this point there are no locks which we care about, we don't keep the identity lock for the entire duration of processing. If we need to do that, it gets a little complicated. I would have to keep a pointer to the locked resource with each event with how the API is currently. Another alternative would be to update the persisted timestamps at this point, but to do that we would want a lock to prevent reads on these processing events while we're updating the timestamps.

I think it's just simpler to use the timeout in memory for now which is not perfect but should be good enough to keep away duplicates and still handle failure.

Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Understood, but I think this is more complicated that it needs to be.

You can lock using entity id, keep the lock, and unlock when processing is done. You don't need to keep a locked resource, because you know the entity id.

All the additional complexity is trying to prevent an error. A much simpler approach would be to process it as simple as possible, and deal with the error. There is no sure way to prevent error.

File: lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java (hunk 11 above)
Comment (alechenninger) [inline lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java]: Rate is num in batch * interval per batch * threads in cluster right? My thinking is just that larger batches are more efficient since lightblue and the event handler can do more work per network call. You can still scale horizontally as needed and get more total threads.

I'll go with per notification id then, but I'll put randomization for another PR since it will need to be weighted so highest priority notifications/events don't have chance of being ignored too long.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 12 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: > You don't need to keep a locked resource, because you know the entity id.

Right now we don't know the callerId though. I have callerId randomized per call so that multiple thread sharing same instance don't produce duplicate events. The callerId is captured in the LockedResource. An alternative would be to make each thread use their own repository and lock strategy, or require any one node be single threaded. I thought it was less complicated in the end to use timestamps + expiration.

File: lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java (hunk 11 above)
Comment (alechenninger) [inline lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java]: Actually I just realized that weighted randomness doesn't matter too much if you don't grab much more than you will try to lock (in other words, if you are trying to lock everything, then there is no risk of something not being processed of course). I might implement this sooner.

For doc events it's a little different because we are grabbing a larger batch than what we will actually process. You had a suggestion below that would change that though, so I'll think about that.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java (hunk 10 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java]: Just to catch this review up with convo we've had in email and such, I'm now somewhat indifferent on the background pinging and I agree it is not necessary but maintain it may help a little. The LockStrategy interface makes this easy to change, so for now I'm focusing on getting everything else right.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 13 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Opened #12

File: lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java (hunk 11 above)
Comment (alechenninger) [inline lib/src/main/java/org/esbtools/eventhandler/NotificationRepository.java]: Randomized notification locks, for doc events opened #13

**Comment by bserdar** [comment]
A minor point: For both notificationRepository and documenteventrepository:
consider renaming retrieveXXXUpTo(int maxEvents) to
retrieveXXXBatch(), with batch size configured with the
concrete implementation (in repository config). Might make things easier.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+            throw new IllegalArgumentException("Time to live should be greater than auto ping " +
+                    "interval, otherwise the lock will likely be lost.");
+        }
+    }
+
+    @Override
+    public LockedResource blockUntilAcquired(String... resourceIds) throws InterruptedException {
```
Comment (bserdar) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java]: Is this really used? This is unsafe without a timeout

Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java]: Not used any more, planning on removing

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 17 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Opened #15
